### PR TITLE
Disable Use Vent For Vanilla (Port from MoreGamemodes)

### PR DESCRIPTION
**Pull request by Tommy-XL**
It seems to work, but I still don’t understand the structure of the code, and it seems there is an option to block certain ventilations

I also don't yet know how to return use vent again

image

### CODE REVIEW

**Comment by Ultradragon005** [comment]
fr

**Comment by Ultradragon005** [comment]
as ridicolous as it sounds, maybe you can reassign their roletype, by #1182 (make a new branch were u merge those 2 to test).
And see if that maybe resets it.
Unless rabek knows another way

**Comment by Tommy-XL** [comment]
I understand how the code works
In fact, if vens is active for player again, need to send `pc.RpcSetVentInteraction()`

**Comment by Ultradragon005** [comment]
oh okay, I noticed that but didn't think it would work, since the original uses Detoriate and not deserialize

**Comment by Tommy-XL** [comment]
Damn, for some roles need to send an Rpc with a check for activate/blocking vents

**Comment by NikoCat233** [comment]
It is working but part of the code is not quite good.
In vanilla ship status is normally updated with messagewriter instead of serializing data.
This code seems to be changing the data for overall ship status, which is unnecessary.
I will try code a better one later

**Review by NikoCat233 — CHANGES_REQUESTED** [review]
I will try do some changes
Gonna open a pr regarding this late today

**Comment by Rabek009** [comment]
NikoCat233 you don't have to code it. There is already better code in More Gamemodes.

**Comment by Rabek009** [comment]
My code doesn't patch shipstatus. It only patch ventilation system.

**Comment by NikoCat233** [comment]
> My code doesn't patch shipstatus. It only patch ventilation system.

https://github.com/Rabek009/MoreGamemodes/blob/5e2ad5467eb9931a0f4235708a9bc82b7d657b82/Patches/ShipStatusPatch.cs#L123

Obviously it is patching overall ship status serializing

And I suggest you check this change
https://github.com/0xDrMoe/TownofHost-Enhanced/pull/1195/files#diff-f28d7cf36b36f9eedd77ae1497343146c6ec19aa116ee0b8b2bdefce37da437aR636

File: Patches/VentSystemPatch.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+namespace TOHE.Patches;
+
+// Patches here is also activated from ShipStatus.Serialize and IntroCutScene 
+// through Utils.SetAllVentInteractions
+
+[HarmonyPatch(typeof(VentilationSystem), nameof(VentilationSystem.Deteriorate))]
+static class VentSystemDeterioratePatch
+{
+    public static Dictionary<byte, int> LastClosestVent;
```
Comment (NikoCat233) [inline Patches/VentSystemPatch.cs:12]: UnUsed variable. Will recode a better version of serialize.

**Review by NikoCat233 — CHANGES_REQUESTED** [review]

File: Patches/HudPatch.cs [hunk 2]
```diff
@@ -119,7 +119,7 @@ public static void Postfix(HudManager __instance)
                     __instance.KillButton.ToggleVisible(false);
                 }
 
-                bool CanUseVent = player.CanUseImpostorVentButton();
+                bool CanUseVent = player.CanUseVent();
                 __instance.ImpostorVentButton.ToggleVisible(CanUseVent);
```
Comment (Tommy-XL) [inline Patches/HudPatch.cs:123]: Found bug, Impostor vent button will be visible for engineer based (for modded)

Comment (NikoCat233) [reply Patches/HudPatch.cs:123]: it should not be changed here i think

**Comment by Tommy-XL** [comment]
Now need to add `RpcSetVentInteraction` for those roles that temporarily block their vents, for example Arsonist, etc.
I'll do this later

**Pull request merged**